### PR TITLE
Feature: Added a CLI subcommand 'list' to list all the workspaces

### PR DESCRIPTION
**Pull request by karthikayan4u**
## References

This PR is with respect to the issue [#10890](https://github.com/jupyterlab/jupyterlab/issues/10890).
No other PRs available for the same.

## Code changes

Changes are made to [jupyterlab/labapp.py](https://github.com/jupyterlab/jupyterlab/blob/master/jupyterlab/labapp.py) as below,
1). Created a class  **LabWorkspaceListApp** within which all the workspaces are collected and displayed with the help of helper functions below.
2). Added a helper function **_list_available_workspaces** which returns a list comprising of all available workspaces.
3). Added another helper function **_load_with_file_times** which will return the respective workspace's **json** from the disk with its stats overwritten accordingly.
4). Made changes to class **LabWorkspaceApp** in such a way that it can handle and call the above class when **list** subcommand is used.

### CODE REVIEW

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/introduce-yourself/17/265

**Comment by baggiponte** [comment]
Not the best place to add this here (I should just add this below #10890) but it helps with my last reply.

How about adding a "metadata.projectRoot" element to the `jupyterlab-workspace`? The value would be the directory where the workspace file is saved to. Then this element might be fetched by `jupyterlab workspace list` as well.

**Comment by agoose77** [comment]
> Not the best place to add this here (I should just add this below #10890) but it helps with my last reply.
> 
> How about adding a "metadata.projectRoot" element to the `*.jupyterlab-workspace`? The value would be the directory where the workspace file is saved to. Then this element might be fetched by `jupyterlab workspace list` as well.

It sounds like you want to customize where you store workspaces. You can do this with `jupyter lab --LabServerApp.workspaces_dir=...`. 
Otherwise, I'm not quite sure what you're asking. In any case, I think it's probably out of scope for this command.

@krassowski what do you think of this PR?

**Comment by baggiponte** [comment]
> It sounds like you want to customize where you store workspaces. You can do this with `jupyter lab --LabServerApp.workspaces_dir=...`.
> Otherwise, I'm not quite sure what you're asking. In any case, I think it's probably out of scope for this command.

Sorry if I was not being clear, I was further elaborating on this point:

> @baggiponte what did you mean by "location"? Did you mean location on disk, or the URL?

I am not thinking of the directory where workspaces are stored. If I understood correctly, workspaces are stored in $JUPYTER_CONFIG_DIR/lab/workspaces and we can also save them in our project directory, as seen below.

image

 What I was referring to was storing the project directory information in the workspace file, so that we are brought in the right directory when workspaces are opeedn. In other words, the lab navigates to the right url where the project is - like with vscode workspaces and Rstudio Rproj files.

This cannot be done with this command - but would eventually be some useful info to display:

```bash
$ jupyter lab workspaces list
my-awesome-proj   /path/to/my-awesome-proj
```

**Comment by agoose77** [comment]
OK, I understand what you're looking to achieve. Workspaces are more like "sessions" that you can distinguish from one another. As such, they record the last directory that was active in the file manager. Is this behaviour not sufficient?

File: jupyterlab/labapp.py [hunk 1]
```diff
@@ -258,6 +260,91 @@ def start(self):
...
+        directory = app.workspaces_dir
+
+        items = [item
+                for item in os.listdir(directory)
+                if item.endswith(WORKSPACE_EXTENSION)]
+        items.sort()
+
+        workspaces = _list_available_workspaces(directory, items)
+        if self.jsonlist:
+            print(json.dumps(workspaces))
```
Comment (bollwyvl) [inline jupyterlab/labapp.py:312]: would want to use `encoding='utf-8'`, probably

File: jupyterlab/labapp.py [hunk 2]
```diff
@@ -258,6 +260,91 @@ def start(self):
...
+            {"LabWorkspaceListApp": {"jsonlist": True}},
+            _i18n("Produce machine-readable JSON list output."),
+        ),
+        json=(
+            {"LabWorkspaceListApp": {"json": True}},
+            _i18n("Produce machine-readable JSON object on each line of output."),
+        ),
+    )
+
+    jsonlist = Bool(
```
Comment (bollwyvl) [inline jupyterlab/labapp.py:283]: perhaps `jsonl` or `jsonlines`, and reference https://jsonlines.org so the format is _slightly_ more specified

**Comment by bollwyvl** [comment]
On the main, my gut says we should be trying to add _less_ in `labapp`. And indeed, as this primarily works with primitives from `jupyterlab_server`, we may want to move 90% of this implementation over there and/or encapsulate the opinions (e.g. encoding, sorting, etc) as a `WorkspaceManager ` over there. This would _still_ require a PR here, once that got merge usptream, as was needed when we added the licenses CLI/API.

File: jupyterlab/labapp.py [hunk 3]
```diff
@@ -14,6 +14,8 @@
 from jupyter_server._version import version_info as jpserver_version_info
 from jupyter_server.serverapp import flags
 from jupyter_server.utils import url_path_join as ujoin
+from jupyter_server import _tz as tz
+from jupyter_server.transutils import _i18n
```
Comment (fcollonval) [inline jupyterlab/labapp.py:18]: JupyterLab is not using the `jupyter_server` translation tool. Could you remove this?

**Review by fcollonval — CHANGES_REQUESTED** [review]

**Comment by krassowski** [comment]
As a general +1 for having this in: it would have helped to debug #11265 a lot. @karthikayan4u would you like to give it one more push to get it through the finish line? Do you need any help in addressing the review?

File: jupyterlab/labapp.py (hunk 1 above)
Comment (fcollonval) [reply jupyterlab/labapp.py:312]: It does not seem to be supported: https://docs.python.org/3/library/json.html#json.dumps

**Pull request closed without merging**